### PR TITLE
added BookController and Mapper

### PR DESCRIPTION
**Pull request by vladshender**

### CODE REVIEW

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 1]
```diff
@@ -10,18 +13,30 @@
...
+        return bookMapper.toDto(bookRepository.save(book));
     }
 
     @Override
-    public List<Book> findAll() {
-        return bookRepository.findAll();
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto)
+                .toList();
```
Comment (sarakhmen) [inline src/main/java/com/example/service/BookServiceImpl.java:40]: is it possible to map list of bookdtos directly using mapper? I mean you shouldn't use stream here and map each dto separately

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

File: pom.xml [hunk 2]
```diff
@@ -22,7 +22,11 @@
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-data-jpa</artifactId>
 		</dependency>
-
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
+			<version>3.3.1</version>
```
Comment (Elena-Bruyako) [inline pom.xml:28]:
```suggestion
```

File: pom.xml [hunk 3]
```diff
@@ -39,6 +43,11 @@
 			<artifactId>mysql-connector-j</artifactId>
 			<scope>runtime</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct</artifactId>
+			<version>1.5.5.Final</version>
```
Comment (Elena-Bruyako) [inline pom.xml:49]: Add <mapstruct.version> to properties 
```suggestion
			<version>${mapstruct.version}</version>
```

File: pom.xml [hunk 4]
```diff
@@ -64,6 +73,32 @@
...
+				<artifactId>maven-compiler-plugin</artifactId>
+				<version>3.11.0</version>
+				<configuration>
+					<source>17</source>
+					<target>17</target>
+					<annotationProcessorPaths>
+						<path>
+							<groupId>org.projectlombok</groupId>
+							<artifactId>lombok</artifactId>
+							<version>1.18.24</version>
```
Comment (Elena-Bruyako) [inline pom.xml:87]:
```suggestion
							<version>${lombok.version}</version>
```

File: pom.xml [hunk 5]
```diff
@@ -64,6 +73,32 @@
 					</excludes>
 				</configuration>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-compiler-plugin</artifactId>
+				<version>3.11.0</version>
+				<configuration>
+					<source>17</source>
```
Comment (Elena-Bruyako) [inline pom.xml:81]: Create <java.version>17</java.version> property and use it

File: src/main/java/com/example/controller/BookController.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/api/books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/BookController.java:17]:
```suggestion
@RequestMapping("/books")
```
Add `server.servlet.context-path=/api` to application.properties

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 7]
```diff
@@ -10,18 +13,28 @@
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
 
+    private final BookMapper bookMapper;
+
     @Autowired
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/BookServiceImpl.java:18]: Remove constructor and use `@RequiredArgsConstructor` instead

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! let's improve your solution!

File: pom.xml (hunk 3 above)
Comment (sarakhmen) [reply pom.xml:51]: not fixed

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Some of the previous mentor's comments were not fixed

File: src/main/java/com/example/controller/BookController.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/BookController.java:17]:
```suggestion
@RequestMapping("/books")
```

**Review by Elena-Bruyako — APPROVED** [review]
Good job

File: src/main/java/com/example/model/Book.java [hunk 9]
```diff
@@ -9,23 +9,37 @@
...
 import lombok.Getter;
 import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
 @Entity
 @Getter
 @Setter
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/Book.java:19]: Where is deprecated, use SQLRestriction instead

File: src/main/java/com/example/repository/SpecificationBuilder.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package com.example.repository;
+
+import com.example.dto.BookSearchParameters;
+import org.springframework.data.jpa.domain.Specification;
+
+public interface SpecificationBuilder<T> {
+    Specification<T> build(BookSearchParameters searchParameters);
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/repository/SpecificationBuilder.java:8]:
```suggestion
```

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 11]
```diff
@@ -1,27 +1,62 @@
...
+    public void deletedById(Long id) {
+        bookRepository.deleteById(id);
     }
 
     @Override
-    public Book save(Book book) {
-        return bookRepository.save(book);
+    @Transactional
+    public void updateBookDetails(Long id, String author, String title, String description) {
+        bookRepository.updateBookDetails(id, author, title, description);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/BookServiceImpl.java:50]:
```suggestion
        Book book = bookRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(.......
        bookMapper.updateBookFromDto(bookDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 12]
```diff
@@ -1,27 +1,62 @@
...
+    @Override
+    public void deletedById(Long id) {
+        bookRepository.deleteById(id);
     }
 
     @Override
-    public Book save(Book book) {
-        return bookRepository.save(book);
+    @Transactional
+    public void updateBookDetails(Long id, String author, String title, String description) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/BookServiceImpl.java:49]:
```suggestion
    public void updateBookDetails(Long id, CreateBookRequestDto bookDto) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments
Please, ask your question in chat

File: src/main/java/com/example/dto/UpdateBookDto.java [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package com.example.dto;
+
+import lombok.Data;
+
+@Data
+public class UpdateBookDto {
+    private String title;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/UpdateBookDto.java:7]: according to the task description Update book request should has more fields

File: src/main/java/com/example/controller/BookController.java (hunk 8 above)
Comment (Elena-Bruyako) [reply src/main/java/com/example/controller/BookController.java:22]: Can't you remove value?

File: src/main/java/com/example/model/Book.java [hunk 14]
```diff
@@ -9,23 +9,37 @@
...
+
     @Column(nullable = false)
     private BigDecimal price;
+
     private String description;
+
     private String coverImage;
+
+    @Column(nullable = false)
+    private Boolean isDeleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/Book.java:44]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 15]
```diff
@@ -1,27 +1,53 @@
...
+    public void deletedById(Long id) {
+        bookRepository.deleteById(id);
     }
 
     @Override
-    public List<Book> findAll() {
-        return bookRepository.findAll();
+    @Transactional
+    public void updateBookDetails(Long id, UpdateBookDto updateBookDto) {
+        Book book = bookRepository.findById(id).stream()
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/BookServiceImpl.java:47]: Why do you use stream here?

File: src/main/resources/db/changelog/changes/02-add-column-is_deleted-in-books.yaml [hunk 16]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-new-column-to-books
+      author: mateacademy
+      changes:
+        - addColumn:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-add-column-is_deleted-in-books.yaml:6]: auto rollback not supports such changes, so in prod application it won't work 
add is_deleted in exists changelog 01

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

**Comment by vladshender** [comment]
thank you for your patience!

File: src/main/java/com/example/dto/UpdateBookDto.java [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package com.example.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class UpdateBookDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/UpdateBookDto.java:7]:
```suggestion
public class UpdateBookRequestDto {
```

File: src/main/java/com/example/dto/UpdateBookDto.java (hunk 13 above)
Comment (Elena-Bruyako) [reply src/main/java/com/example/dto/UpdateBookDto.java:8]: Not fixed
Example of request body
```
{
  "title": "Updated Title",
  "author": "Updated Author",
  "isbn": "978-1234567890",
  "price": 19.99,
  "description": "Updated description",
  "coverImage": "https://example.com/updated-cover-image.jpg"
}
```

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 18]
```diff
@@ -1,27 +1,52 @@
...
     private final BookRepository bookRepository;
 
-    @Autowired
-    public BookServiceImpl(BookRepository bookRepository) {
-        this.bookRepository = bookRepository;
+    private final BookMapper bookMapper;
+
+    @Override
+    public BookDto getBookById(Long id) {
+        Book book = bookRepository.getReferenceById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/BookServiceImpl.java:24]: https://www.baeldung.com/spring-data-jpa-getreferencebyid-findbyid-methods

In this case you can use findById

File: src/main/java/com/example/service/BookServiceImpl.java [hunk 19]
```diff
@@ -1,27 +1,52 @@
...
-        return bookRepository.save(book);
+    public void deletedById(Long id) {
+        bookRepository.deleteById(id);
     }
 
     @Override
-    public List<Book> findAll() {
-        return bookRepository.findAll();
+    @Transactional
+    public void updateBookDetails(Long id, UpdateBookDto updateBookDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/BookServiceImpl.java:46]:
```suggestion
    public BookDto updateBookDetails(Long id, UpdateBookDto updateBookDto) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**